core/templates/services/connection-service.service.ts

```diff
@@ -0,0 +1,122 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service to check for the internet connection.
+ */
+
+import { EventEmitter, Injectable } from '@angular/core';
+import { Subscription, timer } from 'rxjs';
+import { delay, retryWhen, switchMap, tap } from 'rxjs/operators';
+// eslint-disable-next-line oppia/disallow-httpclient
```

Comment:
Why is this required?

Reply:
I used this because if we are using httpclient then we should name the file as backend api service but that name does not suit this service that's why used this check. This check is used in two more places. So still should I rename it?

Reply:
In such a case, you are expected to create an additional `connection-backend-api.service.ts` which only handles HTTP requests and returns promises which can be used in this service. Eg. `admin-backend-api.service.ts` and `admin-data.service.ts`.

Reply:
This file's main function is to send the request and check for internet connection and for that also if we are creating separate file then I don't think a need of two separate files. Maybe we can skip this check for this file only. What do you think?

Reply:
I would suggest these following our existing structure:
1. Have server-connection-backend-api.service.ts [We have backend API service for all HTTP requests]
2. Have internet-connectivity.service.ts [To handle all the logics to running timer, having isOnline function, using server-connection-backend-api.service, having observables, etc.]

In general, we are not expected to use the bypass flag!

Reply:
Done.